### PR TITLE
fix min max zero element

### PR DESCRIPTION
**Pull request by MAbdanM**
Fixing: [#34907](https://github.com/pytorch/pytorch/issues/34907)

Test script:
```
import torch
a = torch.rand(0, 4)
print(a.max(1))
print(a.min(1))
```

## Before:
`RuntimeError: cannot perform reduction function max on tensor with no elements because the operation does not have an identity`

## After:
```
torch.return_types.max(
values=tensor([]),
indices=tensor([], dtype=torch.int64))
torch.return_types.max(
values=tensor([]),
indices=tensor([], dtype=torch.int64))
```

### Convert in numpy then back into tensor again:
```
import torch
a = torch.rand(0, 4)
torch.from_numpy(a.max(1).values.numpy())
```

return:
`tensor([])`

### CODE REVIEW

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 1]
```diff
@@ -73,18 +73,19 @@ inline bool _dimreduce_return_trivial(Tensor &result, const Tensor &self,
...
   if (self.numel() == 1 && self.ndimension() == 0) {
     result.resize_({});
     result.fill_(self);
     return true;
-  }
-
-  if (self.numel() == 0) {
+  } else if (self.numel() == 0 && self.ndimension() == 2) {
+    return false;
+  } else if (self.numel() == 0) {
```
Comment (janucaria) [inline aten/src/ATen/native/ReduceOpsUtils.h]: kalau di dalam statement `if` ada statement `return`, fungsi akan selesai jika masuk conditional `if` tersebut, jadi tidak perlu lagi ada `else`.

line 82 pada conditional if `(self.numel() == 0 && self.ndimension() == 2)` dalamnya `return false` (line 83) sama dengan return fungsi terakhir di line 89. Juga, line 82 dan 84 sama-sama check `self.numel() == 0`. Jadi, lebih baik kalau yang diganti hanya conditional di line 84 menjadi `(self.numel() == 0 && self.ndimension() != 2)`

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 2]
```diff
@@ -73,18 +73,19 @@ inline bool _dimreduce_return_trivial(Tensor &result, const Tensor &self,
   return false;
 }
 
-inline bool _dimreduce_return_trivial_no_ident(Tensor &result, const Tensor &self,
-                                               int64_t dim, bool keepdim, const char *fn_name) {
+inline bool _dimreduce_return_trivial_no_ident(Tensor &result, const Tensor &self, 
+                                               const char *fn_name) {
```
Comment (janucaria) [inline aten/src/ATen/native/ReduceOpsUtils.h]: kode sebelumnya ada parameter `int64_t dim` dan `bool keepdim` yang tidak dipakai dalam fungsi. Sebaiknya dibiarkan seperti itu, tidak dihapus. Kalau mau cleaning unused parameter dalam code, mending buat issue atau PR baru.

Comment (MAbdanM) [reply aten/src/ATen/native/ReduceOpsUtils.h]: Ok, i'll change it, thankyou for advise 🙏

File: aten/src/ATen/native/ReduceOpsUtils.h (hunk 1 above)
Comment (MAbdanM) [reply aten/src/ATen/native/ReduceOpsUtils.h]: nice, thank you :ok_hand:

**Comment by yf225** [comment]
@mruberry @fmassa Would any of you like to review this PR?

**Comment by mruberry** [comment]
Happy to review.

Thanks for the PR, @MAbdanM! I'm not sure I understand its logic, however. Can you elaborate on the check your doing? What about tensors with 1 dimension, or 3 dimensions? What happens if I don't specify a dimension when calling max on the tensor?

**Comment by MAbdanM** [comment]
Okk, here are the result:
```
import torch
a = torch.rand(0, 1)

torch = a.max() # not specify a dimension
# output:  RuntimeError: operation does not have an identity.

torch =  a.max(1)
# output:  torch.return_types.max(values=tensor([]),indices=tensor([], dtype=torch.int64))

torch = a.max(3)
# output: IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 3)

torch = a.max(120)
# output: IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 120)
```

**Comment by mruberry** [comment]
> > Happy to review.
> > Thanks for the PR, @MAbdanM! I'm not sure I understand its logic, however. Can you elaborate on the check your doing? In particular:
> > 
> > * What about tensors with 1 dimension, 3 dimensions, or n dimensions?
> > * What happens if I don't specify a dimension when calling max on the tensor?
> 
> thankyou @mruberry , here are the result:
> 
> ```
> import torch
> a = torch.rand(0, 1)
> 
> torch = a.max() # not specify a dimension
> # output:  RuntimeError: operation does not have an identity.
> 
> torch =  a.max(1)
> # output:  torch.return_types.max(values=tensor([]),indices=tensor([], dtype=torch.int64))
> 
> torch = a.max(3)
> # output: IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 3)
> 
> torch = a.max(120)
> # output: IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 120)
> ```

Yes, but what if the indices are valid? That is, what if the tensor a has 3 or 4 or 5 dims? Then what happens when you do `a.max(3)`?

**Comment by MAbdanM** [comment]
https://github.com/pytorch/pytorch/pull/35062#issuecomment-601599762

I've made changes in self.ndimension() [L84](https://github.com/pytorch/pytorch/commit/845673ecd3aa973ccd00c2d048c9dd20c907b2da#diff-60359cb979cbff83404838d4c1286d32)

```
>>> import torch
>>> a = torch.rand(0) 
>>> b = torch.rand(0, 5)
>>> c = torch.rand(0, 5, 10)
>>> d = torch.rand(0, 5, 10, 15)
>>> e = torch.rand(0, 5, 10, 15, 20)
>>> 
>>> a.max() 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: operation does not have an identity.
>>> b.max(1)
torch.return_types.max(
values=tensor([]),
indices=tensor([], dtype=torch.int64))
>>> c.max(1)
torch.return_types.max(
values=tensor([], size=(0, 10)),
indices=tensor([], size=(0, 10), dtype=torch.int64))
>>> d.max(2)
torch.return_types.max(
values=tensor([], size=(0, 5, 15)),
indices=tensor([], size=(0, 5, 15), dtype=torch.int64))
>>> e.max(3)
torch.return_types.max(
values=tensor([], size=(0, 5, 10, 20)),
indices=tensor([], size=(0, 5, 10, 20), dtype=torch.int64))
```
### Then i compare with numpy:
```
>>> import numpy as np
>>> a = np.random.rand(0) 
>>> b = np.random.rand(0, 5)
>>> c = np.random.rand(0, 5, 10)
>>> d = np.random.rand(0, 5, 10, 15)
>>> e = np.random.rand(0, 5, 10, 15, 20)
>>> 
>>> a.max()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nodeflux/.local/lib/python3.6/site-packages/numpy/core/_methods.py", line 30, in _amax
    return umr_maximum(a, axis, None, out, keepdims, initial, where)
ValueError: zero-size array to reduction operation maximum which has no identity
>>> b.max(1)
array([], dtype=float64)
>>> c.max(1)
array([], shape=(0, 10), dtype=float64)
>>> d.max(2)
array([], shape=(0, 5, 15), dtype=float64)
>>> e.max(3)
array([], shape=(0, 5, 10, 20), dtype=float64)

```

File: test/test_torch.py [hunk 3]
```diff
@@ -14226,6 +14226,23 @@ def test_min_max_binary_op_nan(self, device, dtype):
             self.assertFalse(torch.isnan(ma[i]), "max(a, b): {}, a: {}, b: {}".format(ma[i], a[i], b[i]))
             self.assertFalse(torch.isnan(mi[i]), "min(a, b): {}, a: {}, b: {}".format(mi[i], a[i], b[i]))
 
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float, torch.double)(
+    def test_min_max_empty_element(self, device, dtype):
+        a = torch.rand(0, dtype=dtype, device=device)
```
Comment (mruberry) [inline test/test_torch.py]: You don't need to create so many tensors, but you need to vary the dimension with size zero and test tensors with multiple size zero dimensions, too. For example, what happens when you create a tensor with the shape [1, 0], not [0, 1]?

File: test/test_torch.py [hunk 4]
```diff
@@ -14226,6 +14226,23 @@ def test_min_max_binary_op_nan(self, device, dtype):
...
+        b = torch.rand(0, 5, dtype=dtype, device=device)
+        c = torch.rand(0, 5, 10, dtype=dtype, device=device)
+        d = torch.rand(0, 5, 10, 15, dtype=dtype, device=device)
+        e = torch.rand(0, 5, 10, 15, 20, dtype=dtype, device=device)
+
+        ident_err = 'operation does not have an identity'
+        self.assertRaisesRegex(RuntimeError, ident_err, a.max())
```
Comment (mruberry) [inline test/test_torch.py]: There are other cases where you should get a RunTimeError that you need to check. The following NumPy snippet may suggest some more test cases:

```
a = np.random.rand(1, 0) 
a.max(0)
a.max(1)
```

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 5]
```diff
@@ -81,7 +81,7 @@ inline bool _dimreduce_return_trivial_no_ident(Tensor &result, const Tensor &sel
     return true;
   }
 
-  if (self.numel() == 0) {
+  if (self.numel() == 0 && self.ndimension() < 2) {
```
Comment (oguzserbetci) [inline aten/src/ATen/native/ReduceOpsUtils.h:84]: I think the check of 0 elements should be on the dim. This way it only fails for one dimensional tensors with zero element, or?

**Comment by oguzserbetci** [comment]
I have an improvement over this code. Should I make a new PR? This issue also applies to argmax and argmin #28380, which I also fixed.

Please let me know what is the correct way to provide the code I prepared?

**Comment by mruberry** [comment]
> I have an improvement over this code. Should I make a new PR? This issue also applies to argmax and argmin #28380, which I also fixed.
> 
> Please let me know what is the correct way to provide the code I prepared?

If the changes are related you can include them in this PR. If they're unrelated they should go into a separate PR.

**Comment by oguzserbetci** [comment]
@mruberry I couldn't figure out how to add my commits into this PR. So I opened another #38238. That one tries to fix the problem for all/more reduction functions.

**Pull request closed without merging**